### DISCUSSION TITLE
Layer Activation maps

### DISCUSSION THREAD

**Discussion by someuser2026** [ultralytics/ultralytics#21385] [Q&A] [answered]
Hi, Is there a way to get and plot intermediate activation maps from the trained models for better visualization and debugging?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @someuser2026, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖. An Ultralytics engineer will also assist you here soon!

**Reply by glenn-jocher** [reply]
Yes, you can visualize intermediate activation maps using the built-in `visualize` parameter in the predict function. When you set `visualize=True` or pass a directory path to `visualize`, the model will save feature maps during inference for debugging and visualization purposes.

```python
from ultralytics import YOLO

model = YOLO('yolo11n.pt')
results = model.predict('image.jpg', visualize=True)  # Saves feature maps to runs/detect/predict/
```

You can also specify a custom directory by passing a path to the `visualize` parameter. This will generate feature maps from intermediate layers which you can then analyze to understand what the model is learning at different stages.

**Reply by someuser2026** [reply]
Hi, Thank you! Appreciate the help!

The visualize argument works, however saving to a custom dir does not. The code for the BaseModel also does not appear to accept any argument to specify the custom save dir. It always saves the results in run/task/img_name/layer_or_stage_index.png.

### Code for predict in base model
```
    def predict(self, x, profile=False, visualize=False, augment=False, embed=None):
        """
        Perform a forward pass through the network.

        Args:
            x (torch.Tensor): The input tensor to the model.
            profile (bool): Print the computation time of each layer if True.
            visualize (bool): Save the feature maps of the model if True.
            augment (bool): Augment image during prediction.
            embed (list, optional): A list of feature vectors/embeddings to return.

        Returns:
            (torch.Tensor): The last output of the model.
        """
        if augment:
            return self._predict_augment(x)
        return self._predict_once(x, profile, visualize, embed)

def _predict_once(self, x, profile=False, visualize=False, embed=None):
        """
        Perform a forward pass through the network.

        Args:
            x (torch.Tensor): The input tensor to the model.
            profile (bool): Print the computation time of each layer if True.
            visualize (bool): Save the feature maps of the model if True.
            embed (list, optional): A list of feature vectors/embeddings to return.

        Returns:
            (torch.Tensor): The last output of the model.
        """
        y, dt, embeddings = [], [], []  # outputs
        embed = frozenset(embed) if embed is not None else {-1}
        max_idx = max(embed)
        for m in self.model:
            if m.f != -1:  # if not from previous layer
                x = y[m.f] if isinstance(m.f, int) else [x if j == -1 else y[j] for j in m.f]  # from earlier layers
            if profile:
                self._profile_one_layer(m, x, dt)
            x = m(x)  # run
            y.append(x if m.i in self.save else None)  # save output
            if visualize:
                feature_visualization(x, m.type, m.i, save_dir=visualize)
            if m.i in embed:
                embeddings.append(torch.nn.functional.adaptive_avg_pool2d(x, (1, 1)).squeeze(-1).squeeze(-1))  # flatten
                if m.i == max_idx:
                    return torch.unbind(torch.cat(embeddings, 1), dim=0)
        return x
```

**Reply by glenn-jocher** [reply, marked as answer]
You're correct about the BaseModel implementation, but you're likely calling the high-level `YOLO.predict()` method rather than the BaseModel's `predict()` method directly. The high-level predict method uses different parameters for controlling save locations.

For the high-level predict method, use the `project` and `name` parameters to control where visualizations are saved:

```python
from ultralytics import YOLO

model = YOLO('yolo11n.pt')
results = model.predict('image.jpg', visualize=True, project='custom_dir', name='my_experiment')
```

This will save feature maps to `custom_dir/my_experiment/` instead of the default `runs/detect/predict/` location. The `visualize` parameter at the high-level only accepts boolean values, while the custom directory is controlled through the `project` and `name` parameters.

**Reply by someuser2026** [reply]
Hi, Thank you! It worked!

**Reply by glenn-jocher** [reply]
You're welcome! Glad I could help you get the feature visualization working with custom directories. Feel free to reach out if you have any other questions about YOLO11 or Ultralytics features.

**Reply by someuser2026** [reply]
Hi, just out of curosity, can we also get the gradients for something like GradCAM?

**Reply by glenn-jocher** [reply]
Ultralytics doesn't have built-in GradCAM functionality, but you can implement it yourself using PyTorch's gradient computation. You'll need to register hooks on the target layers and compute gradients with respect to your target class. Since this requires custom implementation beyond the standard predict workflow, you'd typically work with the model's PyTorch components directly rather than through the high-level YOLO interface.

**Reply by someuser2026** [reply]
Hi, appreciate the help. one more question, for the activation maps it spits out quite a few images for each layer. Is there any guide to know what do they actually reperesent? Are they some randomly sampledfeature maps or some specific ones or channel averaged maps, which should result in only one per layer? Appreciate the help!

**Reply by someuser2026** [reply]
Hi, for normal activation maps, is there a way to extract them and use to overlay on the image like heatmaps rather than simply saving as an image?

**Reply by someuser2026** [reply]
Solved. The visualiz automatically also returns the feature maps in .npy files.

**Reply by glenn-jocher** [reply]
Great that you found the .npy files! To answer your earlier question about the multiple images per layer - each image represents a different feature channel from that layer. CNNs produce multiple feature maps per layer (one for each filter/channel), so you're seeing the individual channel activations rather than averaged maps. Early layers typically show edge and texture detectors, while deeper layers capture more complex patterns. The .npy files give you the raw activation data which is perfect for creating custom heatmap overlays.

**Reply by someuser2026** [reply]
Thank you! For the feature maps or indivisual channel activations, how are the ones in the png selected? Generally there are more than 100 per layer but the png files show at most 20. Is it based on the ones with highest pixels being activated?

**Reply by glenn-jocher** [reply]
The selection criteria for which feature maps get saved as PNGs isn't detailed in the current documentation. Typically this could be based on highest activation values, random sampling, or the first N channels, but without examining the specific `feature_visualization` function implementation, I can't say definitively which approach Ultralytics uses. You could check the source code in `ultralytics/utils/plotting.py` or similar visualization utilities to see the exact selection logic, or experiment by comparing the .npy files with the PNG outputs to determine the pattern.
